### ISSUE TITLE
TFA direct call metadata missing in some simple (?) direct calls

### ISSUE THREAD

**Issue by osa1** [dart-lang/sdk#53921]
Repro:

- Clone https://github.com/osa1/typed_array_bench.
- Build SDK with CL https://dart-review.googlesource.com/c/sdk/+/328920.
- Compile benchmark with `$DART_SDK/bin/dart2wasm --dump-kernel-after-tfa test.tfa --js-compatibility test.dart test.wasm`

When I look at the generated code (also confirmed with debug prints, see below), I see that some of the `~/` calls where both arguments are `int`s are not in the direct call metadata.

Some examples are:

- https://github.com/osa1/sdk/blob/d68fa55ea49f6f6371824163d78614a82750650b/sdk/lib/_internal/wasm/lib/js_typed_array.dart#L859
- https://github.com/osa1/sdk/blob/d68fa55ea49f6f6371824163d78614a82750650b/sdk/lib/_internal/wasm/lib/js_typed_array.dart#L487
- https://github.com/osa1/sdk/blob/d68fa55ea49f6f6371824163d78614a82750650b/sdk/lib/_internal/wasm/lib/js_typed_array.dart#L431

Another way to check this with the following change:

```diff
diff --git a/pkg/dart2wasm/lib/code_generator.dart b/pkg/dart2wasm/lib/code_generator.dart
index e813c68c60a..b4e0c4e7fa0 100644
--- a/pkg/dart2wasm/lib/code_generator.dart
+++ b/pkg/dart2wasm/lib/code_generator.dart
@@ -2024,6 +2024,10 @@ class CodeGenerator extends ExpressionVisitor1<w.ValueType, w.ValueType>
       _VirtualCallKind kind,
       void pushReceiver(w.FunctionType signature),
       void pushArguments(w.FunctionType signature)) {
+    if (interfaceTarget.name.toString().contains("~/")) {
+      print("Indirect call to ~/ in ${node.location}");
+    }
+
     SelectorInfo selector = translator.dispatchTable.selectorForTarget(
         interfaceTarget.referenceAs(
             getter: kind.isGetter, setter: kind.isSetter));
```

Outputs look like:

```
Indirect call to ~/ in org-dartlang-sdk:///lib/_internal/wasm/lib/js_typed_array.dart:543:35
Indirect call to ~/ in org-dartlang-sdk:///lib/_internal/wasm/lib/js_typed_array.dart:934:35
Indirect call to ~/ in org-dartlang-sdk:///lib/_internal/wasm/lib/js_typed_array.dart:859:35
Indirect call to ~/ in org-dartlang-sdk:///lib/_internal/wasm/lib/js_typed_array.dart:596:35
Indirect call to ~/ in org-dartlang-sdk:///lib/_internal/wasm/lib/js_typed_array.dart:431:35
Indirect call to ~/ in org-dartlang-sdk:///lib/_internal/wasm/lib/js_typed_array.dart:487:35
Indirect call to ~/ in org-dartlang-sdk:///lib/_internal/wasm/lib/js_typed_array.dart:859:35
...
```

The kernel for one of these methods look like this:

```
  final class JSFloat32ArrayImpl extends _js2::JSFloatArrayImpl implements typ::Float32List { // from org-dartlang-sdk:///lib/_internal/wasm/lib/js_typed_array.dart
    ...
[@vm.procedure-attributes.metadata=methodOrSetterCalledDynamically:false,getterCalledDynamically:false,hasThisUses:false,hasNonThisUses:false,hasTearOffUses:false,getterSelectorId:39]    @#C150
    get length() → core::int
      return [@vm.inferred-type.metadata=dart.core::_BoxedInt] [@vm.direct-call.metadata=dart._js_types::JSFloat32ArrayImpl.lengthInBytes] [@vm.inferred-type.metadata=!] this.{_js2::JSFloat32ArrayImpl::lengthInBytes}{core::int}.{core::num::~/}(4){(core::num) → core::int};
    ...
  }
```

If I'm reading this right, the direct call metadata includes one of the calls in the method (`JSFloat32ArrayImpl.lengthInBytes`), but the `~/` call is not direct.

**Comment by osa1** [comment]
Note: overriding `Target.concreteIntLiteralClass` in dart2wasm target implementation (https://dart-review.googlesource.com/c/sdk/+/333004) makes no difference in this benchmark.

**Comment by alexmarkov** [comment]
TFA doesn't treat `int` class specially. So the type of int * int is calculated using `operator *` in`_BoxedInt` which is declared in the following way: 

```dart
  external num operator *(num other);
```

As a result, inferred type of `JSFloat32ArrayImpl.lengthInBytes` (which is a result of multiplication) is `{_BoxedDouble, _BoxedInt}`.

Consequently, a call to `operator ~/` has two possible targets and it is not devirtualized.

```dart
  @override
  @pragma("wasm:prefer-inline")
  int get lengthInBytes => (super.lengthInBytes ~/ 4) * 4;

  @override
  @pragma("wasm:prefer-inline")
  int get length => lengthInBytes ~/ 4;
```

**Comment by osa1** [comment]
We discussed this a little bit more in chat. The problem is in `length`, type of `lengthInBytes` from TFA's point of view is not `int`, it's `{_BoxedDouble, _BoxedInt}`.

It's unclear to me how can the TFA type be less precise than the static type of `lengthInBytes`, however @alexmarkov says if we implement `num` operators with double dispatch (as in VM) it should do the trick.

So instead of `extern operator *(num arg)`, we do something like

```dart
class _BoxedInt {
  num operator *(num right) {
    right._multiplyInt(this);
  }

  external int _multiplyInt(int left);
}
```

**Comment by alexmarkov** [comment]
Small standalone repro:

```dart
var z;
void main() {
  z = 2 * 3;
  print(z ~/ 2);
}
```

Kernel after TFA:
```
  static field dynamic z;
  static method main() → void {
    foo::z = [@vm.direct-call.metadata=dart.core::_BoxedInt.*] [@vm.inferred-type.metadata=! (skip check)] 2.{core::num::*}(3){(core::num) → core::int};
    core::print([@vm.inferred-type.metadata=dart.core::_BoxedInt] foo::z{dynamic}.~/(2));
  }
```

I think we have 2 options of fixing this problem:

- Add special case for int binary operations to TFA (similarly to a special case in the Dart static type system) so it would infer that `int * int` has a type `int` regardless of the implementation of `int.operator *`.
- Provide an implementation of `int.operator *` in dart2wasm core library which would result in `int` if its argument is `int`, and avoid such special cases in TFA. This would be similar to the VM/AOT setup.

If the second approach turns out to be problematic for some reason, I can help with the first approach.

**Comment by mkustermann** [comment]
If I understand correctly: If there's a `int * int` callsite where the `*` operation on the integer class is declared as `external num operator*(num)` , then TFA assumes any `num` can be returned so it concludes `int * int => num`. It would then propagate `num` around - which seems not good.

The static type system is more precise on the `int * int` expression as it knows it's an `int`. 

@alexmarkov When TFA builds the summaries it could ensure to emit `Narrow(<return-value>, <static-return-type>)` nodes (iff the interface return type  is different from the static expression type) thereby making this more precise. It would be general mechanism, not specific to integers per se. As summary building isn't contributing so much to the TFA runtime, are there any reasons not to do that?

**Comment by alexmarkov** [comment]
@mkustermann What you suggest effectively works only for int operations in dart2wasm, so this is just another way of implementing the first approach (note that returning a type parameter is already handled in TFA). But I like your suggestion as we won't need to recognize and hardcode binary int operations, so I gave it a try.

As you mention, the cost of building summaries is insignificant, but the number of additional narrowing operations during interpretation of a summary is important (as we may need to "run" summary many times during the analysis). I did a few experiments:
* On a large Flutter app built with AOT, narrowing all call sites by their static types increases compilation time by 2.5%, while size of the resulting AOT snapshot stays the same.
* Narrowing call sites only if static type is different from the return type of a target doesn't regress compilation time (and also doesn't make any difference in the AOT snapshot size). So this approach looks feasible.

However, after improving inference in TFA I got a bunch of new dart2wasm test failures: https://dart-review.googlesource.com/c/sdk/+/333580. Investigating them now.
